src/IO/S3/Credentials.cpp

```diff
@@ -233,8 +226,78 @@ std::shared_ptr<AWSEC2MetadataClient> InitEC2MetadataClient(const Aws::Client::C
+    return ec2_metadata_service_endpoint;
+}
+
+std::shared_ptr<AWSEC2MetadataClient> InitEC2MetadataClient(const Aws::Client::ClientConfiguration & client_configuration)
+{
+    auto endpoint = awsMetadataEndpoint();
+    return std::make_shared<AWSEC2MetadataClient>(client_configuration, endpoint.c_str());
+}
+
+String AWSEC2MetadataClient::getAvailabilityZoneOrException()
+{
+    Poco::URI uri(awsMetadataEndpoint() + EC2_AVAILABILITY_ZONE_RESOURCE);
+    Poco::Net::HTTPClientSession session(uri.getHost(), uri.getPort());
+
+    Poco::Net::HTTPResponse response;
+    Poco::Net::HTTPRequest request(Poco::Net::HTTPRequest::HTTP_GET, uri.getPath());
+    session.sendRequest(request);
+
+    std::istream& rs = session.receiveResponse(response);
+    if (response.getStatus() != Poco::Net::HTTPResponse::HTTP_OK) {
+        throw DB::Exception(ErrorCodes::AWS_ERROR,
+            "Failed to get AWS availability zone. HTTP response code: {}", response.getStatus());
+    }
+    String response_data;
+    Poco::StreamCopier::copyToString(rs, response_data);
+    return response_data;
+}
+
+String getGCPAvailabilityZoneOrException()
+{
+    Poco::URI uri("http://169.254.169.254/computeMetadata/v1/instance/zone");
+    Poco::Net::HTTPClientSession session(uri.getHost(), uri.getPort());
+    Poco::Net::HTTPRequest request(Poco::Net::HTTPRequest::HTTP_GET, uri.getPath());
+    Poco::Net::HTTPResponse response;
+    request.set("Metadata-Flavor",  "Google");
+    session.sendRequest(request);
+    std::istream& rs = session.receiveResponse(response);
+    if (response.getStatus() != Poco::Net::HTTPResponse::HTTP_OK) {
+        throw DB::Exception(ErrorCodes::GCP_ERROR,
+            "Failed to get GCP availability zone. HTTP response code: {}", response.getStatus());
+    }
+    String response_data;
+    Poco::StreamCopier::copyToString(rs, response_data);
+    Strings zone_info;
+    boost::split(zone_info, response_data, boost::is_any_of("/"));
+    if (zone_info.size() != 4) {
+        throw DB::Exception(ErrorCodes::GCP_ERROR,
+            "Invalid format of GCP zone information, expect projects/<project-number>/zones/<zone-value>, got {}", response_data);
+    }
+    return zone_info[3];
+}
+
+String getRunningAvailabilityZone()
+{
+    LOG_INFO(&Poco::Logger::get("Application"), "Trying to detect the availability zone.");
+
+    try{
+        auto aws_az = AWSEC2MetadataClient::getAvailabilityZoneOrException();  
+        return aws_az;
+    }
+    catch (const DB::Exception & aws_ex)
+    {
+        try{
+            auto gcp_zone = getGCPAvailabilityZoneOrException();
+            return gcp_zone;
+        }
+        catch(const DB::Exception & gcp_ex)
+        {
+            throw DB::Exception(ErrorCodes::UNSUPPORTED_METHOD,
+            "Failed to find the availability zone, tried AWS, GCP. AWS Error {}\n GCP Error {}", aws_ex.displayText(), gcp_ex.displayText());
+        }
+    }
```

Comment:
Let's make this result static into static variable? Like I write in the comment above.

Reply:
Just realized that after last push. Fixed.